### ISSUE TITLE
protect::function compile error with clang

### ISSUE THREAD

**Issue by romainfrancois** [r-lib/cpp11#87]
``` r
cpp11::cpp_source(code = '
#include <cpp11.hpp>

[[cpp11::register]]
int fun() {
  cpp11::stop("b");
  cpp11::warning("c");
}
', quiet = FALSE)
#> clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I/Users/romainfrancois/.R/library/4.0/cpp11/include  -I/usr/local/include   -fPIC  -Wall -O3 -Wall -Wimplicit-int-float-conversion -c /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpDqea6R/file15b0939b498b1/src/code_0.cpp -o /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpDqea6R/file15b0939b498b1/src/code_0.o
#> In file included from /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpDqea6R/file15b0939b498b1/src/code_0.cpp:2:
#> In file included from /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11.hpp:5:
#> In file included from /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/as.hpp:9:
#> /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/protect.hpp:236:10: error: definition with same mangled name '_ZNK5cpp117protect8functionIFvPKczEEclIJRS3_EEEDTclsr6detailE5applyclL_ZNSt3__17declvalIPS4_EEDTclsr3std3__1E9__declvalIT_ELi0EEEvEEclsr3stdE16forward_as_tuplespclsr3stdE7forwardIT_Efp_EEEEDpOSD_' as another definition
#>     auto operator()(A&&... a) const
#>          ^
#> /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/protect.hpp:236:10: note: previous definition is here
#> 1 error generated.
#> make: *** [/private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpDqea6R/file15b0939b498b1/src/code_0.o] Error 1
#> Error in dyn.load(shared_lib, local = TRUE, now = TRUE): unable to load shared object '/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T//RtmpDqea6R/file15b0939b498b1/src/code_0.so':
#>   dlopen(/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T//RtmpDqea6R/file15b0939b498b1/src/code_0.so, 6): image not found
```

<sup>Created on 2020-08-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

ping @bkietz 

This failed to compile `array.cpp` iin arrow, which looks like: 

```cpp
void arrow::r::validate_slice_length(R_xlen_t length, int64_t available) {
  if (length == NA_INTEGER) {
    cpp11::stop("Slice 'length' cannot be NA");
  }
  if (length < 0) {
    cpp11::stop("Slice 'length' cannot be negative");
  }
  if (length > available) {
    cpp11::warning("Slice 'length' greater than available length");
  }
}
```

@jimhester can we try to fix this before #84 ?

**Comment by jimhester** [comment]
I submitted 0.2.0 on Friday, but we can do another release in the not too distant future.

**Comment by jimhester** [comment]
Also this seems to only be an issue with clang, gcc compiles this without error.

**Comment by jimhester** [comment]
You can workaround the llvm issue using 

```diff
diff --git a/inst/include/cpp11/protect.hpp b/inst/include/cpp11/protect.hpp
index fe5f8aa..bcb6781 100644
--- a/inst/include/cpp11/protect.hpp
+++ b/inst/include/cpp11/protect.hpp
@@ -271,12 +271,12 @@ void stop [[noreturn]] (const std::string& fmt, Args... args) {

 template <typename... Args>
 void warning(const char* fmt, Args... args) {
-  safe[Rf_warning](fmt, args...);
+  unwind_protect([&] { Rf_warning(fmt, args...); });
 }

 template <typename... Args>
 void warning(const std::string& fmt, Args... args) {
-  safe[Rf_warning](fmt.c_str(), args...);
+  unwind_protect([&] { Rf_warning(fmt.c_str(), args...); });
 }

 }  // namespace cpp11
```

But that then doesn't work with gcc 4.8.5

**Comment by romainfrancois** [comment]
That's fine, if anything we can vendor `cpp11` if we need to release it. I'll let @bkietz come up with a compiler conditional compilation strategy as this seems to be a mix of too many things I'm not that confortable with (constexpr, tuples, declval, ...)

**Comment by bkietz** [comment]
The root of the problem is that `Rf_error` and `Rf_warning` don't have identical type because Rf_error carries `[[noreturn]]` *but* the two types do mangle identically
```
<stdin>:35:3: error: static_assert failed due to requirement 'std::is_same<void (const char *, ...) __attribute__((noreturn)), void (const char *, ...)>::value' ""
  static_assert(std::is_same<decltype(Rf_error), decltype(Rf_warning)>::value, "");
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by nealrichardson** [comment]
I don't fully understand why this is a problem now and not before but we can also work around this in `arrow` by removing the warning.

**Comment by jimhester** [comment]
Even if you can work around the particular issue in arrow being able to have a `cpp11::stop()` and a `cpp11::warning()` in the same compilation unit seems something we need to support.

Does removing the `[[noreturn]]` annotation fix this issue then?

**Comment by romainfrancois** [comment]
And/or can we embed the `no-return-ness` in the type somehow ?

**Comment by bkietz** [comment]
fixes the issue for me: https://github.com/r-lib/cpp11/pull/85/commits/25014d84752fe5f237a43a7984b313f2647ba7ab

**Comment by bkietz** [comment]
> Does removing the [[noreturn]] annotation fix this issue then?

Yes, but I don't think that will be necessary.

> And/or can we embed the no-return-ness in the type somehow ?

I tried that, but clang and gcc don't handle attributes consistently enough to write `enable_if_noreturn` (or at least I couldn't swing it).

**Comment by bkietz** [comment]
resolved by #85
